## Warm Ups

__Exercise 1.1 (10 points).__ In lecture, we saw that Python's lists have reference semantics. Give a similar example that shows Python's dictionaries have reference semantics. Make sure to explain your example.

In [1]:
#changing the list with one variable changes the list for the other variable.
# Here we generate a dictionary with name and age elements
x = {"name": "John", "Age": 21}
x

{'name': 'John', 'Age': 21}

In [2]:
y = x
# Change the value for Age element in y
y["Age"]= 19
y

{'name': 'John', 'Age': 19}

In [3]:
x

{'name': 'John', 'Age': 19}

**We can see x also has the corresponding change.**
**It proves Python's dictionaries have reference semantics.**

__Exercise 1.2 (10 points).__ Are strings mutable? Give an example and explain your reasoning. _Hint: It's a good idea to confirm your answer by checking the Python 3 documentation._

Next, read [the documentation](https://docs.python.org/3/library/stdtypes.html#str.replace) for the string `.replace()` method. Give an example of using the method and explain how it fits in with your understanding of whether or not strings are mutable.

In [4]:
# Create a string x
x = "Hello"

In [5]:
# check the type of "Hello"
type(x)

str

In [6]:
# change the string
# x[1] = "A"

**String is not mutable.**

In [7]:
x.replace("e","A")

'HAllo'

In [8]:
x = "Hello, world; Hello, world."

In [9]:
print(x.replace("Hello","Bye",1))

Bye, world; Hello, world.


In [12]:
print(x.replace("Hello","Bye",2))

Bye, world; Bye, world.


**A string is not mutable but it can be replaced. Here "Hello" is not changed but replaced by "Bye". If we print x again it would still be the same as before.**

In [226]:
x

'Hello, world; Hello, world.'

## Writing Functions

__Exercise 1.3 (10 points).__ Write a function `mean` that takes a list of numbers as input and returns their mean as output. Test your function on at least two examples.

In [27]:
# write a function
def mean(x):
    a = len(x)  # calculate the number of the values in a list
    b = sum(x)  # calculate the sum of the values in a list
    return b/a  # mean= sum /number of values

In [28]:
x = [1, 3, 5]
mean(x)

3.0

In [29]:
y = [1, 2, 3, 4, 5, 6]
mean(y)

3.5

__Exercise 1.4 (10 points).__ For the function you wrote in Exercise 1.3, what happens if the input list is empty or contains non-numeric elements? Create a new version of your function called `better_mean` that returns `None` when either of these unusual cases occur.

_Hint: A similar problem is discussed in [Section 6.8](http://greenteapress.com/thinkpython2/html/thinkpython2007.html#sec77) of Think Python._

In [120]:
def better_mean(n):
    if not all(isinstance(i, (int, float)) for i in n): 
        #input list contains non-numeric elements')
        return None 
    elif len(n) == 0:
        # input list is empty
        return None 
    else:
        a = len(n)
        b = sum(n)
        return b/a

In [121]:
z = [1, 2.2, 3, 4, 5, 6]
better_mean(z)

3.533333333333333

In [89]:
z = []
better_mean(z)

In [33]:
z = "Hello"
better_mean(z)

__Exercise 1.5 (10 points).__ Read [Section 4.9](http://greenteapress.com/thinkpython2/html/thinkpython2005.html#sec50) of Think Python. Create a new version of your function from Exercise 1.4 called `best_mean` that includes a docstring explaining how to use the function. Give an example to show that your docstring works with Python's built-in help system.

In [74]:
def best_mean(n):
    """Calculate mean for a list of numbers, if the list is not empty and has no non-numeric values."""
    if not all(isinstance(i, (int, float)) for i in n):
        return None       
    elif len(n) == 0:
        return None 
    else:
        a = len(n)
        b = sum(n)
        return b/a

In [75]:
help(best_mean)

Help on function best_mean in module __main__:

best_mean(n)
    Calculate mean for a list of numbers, if the list is not empty and has no non-numeric values.



In [231]:
x = [1, 2, 3, 4, 5, 6]
best_mean(x)

3.5

__Exercise 1.6 (10 points).__ Write a function `median` that takes a list of numbers as input and returns their median as output. Document your function with a docstring. Test your function on at least two examples.

In [76]:
def median(x):
    """calculate median for a list of numbers"""
    n = len(x) #calculate length of a list
    s = sorted(x) #sort the list from samllest to biggest"""
    if n % 2 == 0:
        return s[n//2-1]/2 +s[n//2]/2
        #If the number of values in a list is even, the median is the average of the two middle numbers"""
    else:
        return s[n//2]
        #If the number of values in a list is odd, the median is the middle number"""

In [77]:
x = [1, 2, 3, 4]
median(x)

2.5

In [78]:
y = [1, 2, 3, 4, 5]
median(y)

3

## Finding Exponential Roots

The Newton-Raphson algorithm is an algorithm for finding the zeroes of a mathematical function. We can use the Newton-Raphson algorithm to find zeroes of the function

$$
f(x) = x^p - c
$$

where $p$ and $c$ are constants. For example, if we choose $p = 2$ and $c = 5$, the Newton-Raphson algorithm finds solutions to

$$
0 = x^2 - 5
$$

In other words, we can use the algorithm to find square roots. By changing $p$, we can also find other kinds of exponential roots.


The algorithm works by starting from an initial guess $x_0$ and then iteratively evaluating

$$
x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)}
$$

for $n = 0, 1, 2, \dots, N$ until reaching a result $X_{N+1}$ with acceptable accuracy. The initial guess does not need to be an excellent guess, but can affect which zero is found.

For our specific function $f$, note that $f'(x) = px^{p-1}$.

__Exercise 1.7 (20 points).__ Write a function `root` that uses the Newton-Raphson algorithm to compute one of the $p$-th roots for a constant $c$. Your function does not need to find complex roots, only real roots. Your function should have arguments

* `c`, the constant
* `p`, the power
* `x0`, the initial guess
* `N`, the maximum number of iterations

Test your function for $c = 2$, $p = 2$, $N = 200$. Try different values of $x_0$. Can you find initial guesses to get both roots? Explain what happens when the initial guess is $x_0 = 0$.

In [10]:
def root(c, p, x0, N):
    for i in range (1, N+1):
        f_x = x0 ** p - c       #calculate f(x)
        d_f_x = p * (x0 **(p-1)) #calculate 𝑓′(x)
        a = f_x/ d_f_x 
        x1 = x0 - a #use the formula to calculate next x
        x0 = x1 #replace the 𝑥𝑛 and 𝑥𝑛+1
    print (x0)
    

In [11]:
root(2,2,2,200)

1.414213562373095


In [12]:
root(2,2,-2,200)

-1.414213562373095


**When the initial guess is 2, the root is 1.414213562373095; when the initial guess is -2, the root is -1.414213562373095.**

In [13]:
# root(2,2,0,200)

**If the initial guess x0 is 0, 𝑓′(𝑥)= p * x0^(p-1) will be 0 in the denominater to calculate x1. The denominator of any fraction cannot have the value zero.**

__Exercise 1.8 (20 points).__ Read the Python documentation for the string `.format()` method (see [here](https://docs.python.org/3/library/stdtypes.html#str.format)). Create a new version of your function from Exercise 1.7 called `root_print` that neatly prints the iteration number and estimate for each iteration up to $N$.

Test your function for $c = 7$, $p = 3$, $x_0 = 2$, $N = 50$. About how many iterations does it take for the first 3 digits to stabilize at the correct values?

In [1]:
def root_print(c, p, x0, N):
    for i in range (1, N+1):
        f_x = x0 ** p - c       #calculate f(x)
        d_f_x = p * (x0 **(p-1)) #calculate 𝑓′(x)
        a = f_x/ d_f_x 
        x1 = x0 - a #use the formula to calculate next x
        x0 = x1 #replace the 𝑥𝑛 and 𝑥𝑛+1
        w0 = "The iteration is {}, and the root is {}.".format(i, x0)
        print (w0)

In [2]:
root_print(7,3,2,50)

The iteration is 1, and the root is 1.9166666666666667.
The iteration is 2, and the root is 1.9129384583070783.
The iteration is 3, and the root is 1.9129311828000604.
The iteration is 4, and the root is 1.9129311827723892.
The iteration is 5, and the root is 1.9129311827723892.
The iteration is 6, and the root is 1.9129311827723892.
The iteration is 7, and the root is 1.9129311827723892.
The iteration is 8, and the root is 1.9129311827723892.
The iteration is 9, and the root is 1.9129311827723892.
The iteration is 10, and the root is 1.9129311827723892.
The iteration is 11, and the root is 1.9129311827723892.
The iteration is 12, and the root is 1.9129311827723892.
The iteration is 13, and the root is 1.9129311827723892.
The iteration is 14, and the root is 1.9129311827723892.
The iteration is 15, and the root is 1.9129311827723892.
The iteration is 16, and the root is 1.9129311827723892.
The iteration is 17, and the root is 1.9129311827723892.
The iteration is 18, and the root is 1.9

 **It took three iteration for the first 3 digits after decimal (.912) to stabilize at the correct values.**

## Fibonacci Words

A [Fibonacci word](https://en.wikipedia.org/wiki/Fibonacci_word) is a string of 0s and 1s constructed by repeatedly concatenating strings. The first two words are

```
S0 = "0"
S1 = "01"
```

Then each word is formed by concatenating the previous two words in the sequence. For instance, `S2`, is formed by concatenating `S1` and `S0`. So

```
S2 = "010"
S3 = "01001"
```

__Exercise 1.9 (20 points).__ Write a function `fib` that computes Fibonacci words. Your function should take an argument `n` that specifies which word to compute.

Use your function to print the first 10 Fibonacci words.

In [8]:
def fib(n):
    S_0 = "0"
    S_1 = "01"
    a = ""
    for i in range(0, n):
        if i == 0:
            print ("S0 = 0.")
        elif i == 1:
            print ("S1 = 01.")
        else:
            a = S_1
            S_1 += S_0 
        #+= operator lets you add two values together and assign the resultant 
        #value to a variable. combine s0 and s1 and assign the value to s1.
            S_0 = a
            w0 = "S{} = {}.".format(i, S_1)
            print(w0) 
        

In [9]:
fib(10)

S0 = 0.
S1 = 01.
S2 = 010.
S3 = 01001.
S4 = 01001010.
S5 = 0100101001001.
S6 = 010010100100101001010.
S7 = 0100101001001010010100100101001001.
S8 = 0100101001001010010100100101001001010010100100101001010.
S9 = 01001010010010100101001001010010010100101001001010010100100101001001010010100100101001001.
